### PR TITLE
RCT1: import vehicles

### PR DESCRIPTION
**Pull request by IntelOrca**

### CODE REVIEW

File: src/openrct2/rct1/S4Importer.cpp [hunk 1]
```diff
@@ -843,11 +843,16 @@ class S4Importer : public IS4Importer
...
-            vehicleEntryIndex = 1;
-        }
+//        uint8 vehicleEntryIndex = 0;
+//        if (src->vehicle_type == 36)
+//        {
+//            vehicleEntryIndex = 1;
+//        }
+        char buf[64];
+        sprintf(buf, "%d", src->vehicle_type);
+        log_error(buf);
```
Comment (Gymnasiast) [inline src/openrct2/rct1/S4Importer.cpp]: Ok, this is debugging data, should have left this out

**Comment by Gymnasiast** [comment]
crumbly woods 2017-01-07 00-18-27

**Comment by Gymnasiast** [comment]
The last commit seems to improve the behaviour of guests getting on and off rides - previously, they wouldn't disembark.

File: src/openrct2/rct1/S4Importer.cpp [hunk 2]
```diff
@@ -594,24 +594,30 @@ class S4Importer : public IS4Importer
...
+            dst->last_peep_in_queue[i] = src->last_peep_in_queue[i];
             dst->queue_length[i] = src->num_peeps_in_queue[i];
         }
         dst->num_stations = src->num_stations;
 
-        for (int i = 0; i < 32; i++)
+        // Vehicle links (indexes converted later)
+        for (int i = 0; i < RCT1_MAX_VEHICLES_PER_RIDE; i++)
+        {
+            dst->vehicles[i] = src->vehicles[i];
```
Comment (duncanspumpkin) [inline src/openrct2/rct1/S4Importer.cpp]: Could be wrong here but shouldn't only the train vehicles (i.e. first car) end up in the ride vehicle list.

File: src/openrct2/rct1/S4Importer.cpp [hunk 3]
```diff
@@ -775,46 +798,224 @@ class S4Importer : public IS4Importer
...
             {
-                rct1_peep *srcPeep = &_s4.sprites[i].peep;
-                if (srcPeep->x != (sint16)0x8000 || srcPeep->state == PEEP_STATE_ON_RIDE)
+                rct1_vehicle * srcVehicle = &_s4.sprites[i].vehicle;
+                if (srcVehicle->x != (sint16)0x8000)
                 {
-                    rct_peep *peep = (rct_peep*)create_sprite(SPRITE_IDENTIFIER_PEEP);
-                    move_sprite_to_list((rct_sprite*)peep, SPRITE_LIST_PEEP * 2);
+                    rct_vehicle * vehicle = (rct_vehicle *)create_sprite(SPRITE_IDENTIFIER_VEHICLE);
+                    move_sprite_to_list((rct_sprite *)vehicle, SPRITE_LIST_VEHICLE * 2);
```
Comment (duncanspumpkin) [inline src/openrct2/rct1/S4Importer.cpp]: I think this should only be trains in this list not all vehicles.

**Comment by janisozaur** [comment]
Even with 9c9aa1b, [Luna Park](https://github.com/OpenRCT2/OpenRCT2/issues/3552) still hits the `Unsupported RCT1 colour.` assert for me.

@Gymnasiast said it works for him even on 7ccfc3d, so I'm not sure what's going on. Could this be due to some objects I have installed?

Here's how the stack looks:
```
#4  0x0000000000f1d775 in Guard::Assert_VA (expression=false, message=0x14db720 "Unsupported RCT1 colour.", args=0x7fffffffa2b0) at /home/janisozaur/workspace/OpenRCT2/src/openrct2/core/Guard.cpp:98
        __PRETTY_FUNCTION__ = "void Guard::Assert_VA(bool, const char*, __va_list_tag*)"
#5  0x0000000000f1d6b3 in Guard::Assert (expression=false, message=0x14db720 "Unsupported RCT1 colour.") at /home/janisozaur/workspace/OpenRCT2/src/openrct2/core/Guard.cpp:51
        args = <error reading variable args (Attempt to dereference a generic pointer.)>
#6  0x00000000010f81a3 in Guard::ArgumentInRange<unsigned long> (argument=255, min=0, max=32, message=0x14db720 "Unsupported RCT1 colour.") at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/../core/Guard.hpp:54
        args = <error reading variable args (Attempt to dereference a generic pointer.)>
#7  0x00000000010f3eed in RCT1::GetColour (colour=255 '\377') at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/tables.cpp:70
        map = "\000\001\002\004\005\006\a\t\v\f\r\016\017\020\022\023\024\025\026\027\030\031\032\033\034\036\037\035\003\n\021\b"
#8  0x00000000010d6126 in S4Importer::ImportVehicle (this=0x7fffced23800, dst=0x25af000 <_spriteList>, src=0x7fffced83f24) at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/S4Importer.cpp:966
        buf = <error reading variable buf (Cannot access memory at address 0xffffffffffffffb3)>
        __func__ = "ImportVehicle"
        vehicleEntryIndex = 0 '\000'
        ride = 0x28f4a00 <gRideList+19456>
#9  0x00000000010d1dc6 in S4Importer::ImportVehicles (this=0x7fffced23800) at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/S4Importer.cpp:847
        vehicle = 0x25af000 <_spriteList>
        srcVehicle = 0x7fffced83f24
        i = 7
        vehicles = std::vector of length 1, capacity 1 = {[0] = 0x25af000 <_spriteList>}
        spriteIndexMap = {[0] = 65535, [1] = 65535, [2] = 65535, [3] = 65535, [4] = 65535, [5] = 65535, [6] = 65535, [7] = 0 <repeats 2221 times>, [2228] = 52250, [2229] = 63217, [2230] = 32767, [2231] = 0, [2232] = 63051, [2233] = 35898, [2234] = 0, [2235] = 0, [2236] = 14080, [2237] = 55080, [2238] = 52963, [2239] = 9983, [2240] = 38432, [2241] = 63325, [2242] = 32767, [2243] = 0, [2244] = 19264, [2245] = 63254, [2246] = 32767, [2247] = 0, [2248] = 14264, [2249] = 63480, [2250] = 32767, [2251] = 0, [2252] = 11424, [2253] = 63288, [2254] = 32767, [2255] = 0, [2256] = 51056, [2257] = 5, [2258] = 24928, [2259] = 0, [2260] = 52250, [2261] = 63217, [2262] = 32767, [2263] = 0, [2264] = 63051, [2265] = 35898, [2266] = 24928, [2267] = 0, [2268] = 14080, [2269] = 55080, [2270] = 52963, [2271] = 9983, [2272] = 38432, [2273] = 63325, [2274] = 32767, [2275] = 0, [2276] = 19264, [2277] = 63254, [2278] = 32767, [2279] = 0, [2280] = 14264, [2281] = 63480, [2282] = 32767, [2283] = 0, [2284] = 11424, [2285] = 63288, [2286] = 32767, [2287] = 0, [2288] = 47216, [2289] = 5, [2290] = 24928, [2291] = 0, [2292] = 552, [2293] = 0, [2294] = 0, [2295] = 0, [2296] = 47232, [2297] = 5, [2298] = 24928, [2299] = 0, [2300] = 40811, [2301] = 63206, [2302] = 32767, [2303] = 0, [2304] = 552, [2305] = 0, [2306] = 0, [2307] = 0, [2308] = 47232, [2309] = 5, [2310] = 24928, [2311] = 0, [2312] = 552, [2313] = 0, [2314] = 0, [2315] = 0, [2316] = 11972, [2317] = 63218, [2318] = 32767, [2319] = 0, [2320] = 47232, [2321] = 5, [2322] = 24928, [2323] = 0, [2324] = 53507, [2325] = 63205, [2326] = 32767, [2327] = 0, [2328] = 0, [2329] = 0, [2330] = 0, [2331] = 0, [2332] = 47361, [2333] = 65535, [2334] = 32767, [2335] = 0, [2336] = 49664, [2337] = 65535, [2338] = 32767, [2339] = 0, [2340] = 12288, [2341] = 63484, [2342] = 32767, [2343] = 0, [2344] = 552, [2345] = 0, [2346] = 0, [2347] = 0, [2348] = 1, [2349] = 0, [2350] = 0, [2351] = 0, [2352] = 61025, [2353] = 63215, [2354] = 32767, [2355] = 0, [2356] = 14080, [2357] = 55080, [2358] = 52963, [2359] = 9983, [2360] = 0, [2361] = 0, [2362] = 0, [2363] = 0, [2364] = 47600, [2365] = 65535, [2366] = 32767, [2367] = 0, [2368] = 49728, [2369] = 65535, [2370] = 32767, [2371] = 0, [2372] = 12288, [2373] = 63484, [2374] = 32767, [2375] = 0, [2376] = 552, [2377] = 0, [2378] = 0, [2379] = 0, [2380] = 1, [2381] = 0, [2382] = 0, [2383] = 0, [2384] = 61025, [2385] = 63215, [2386] = 32767, [2387] = 0, [2388] = 60965, [2389] = 63215, [2390] = 32767, [2391] = 0, [2392] = 0, [2393] = 0, [2394] = 0, [2395] = 0, [2396] = 0, [2397] = 0, [2398] = 30, [2399] = 0, [2400] = 47616, [2401] = 65535, [2402] = 32767, [2403] = 0, [2404] = 2, [2405] = 0, [2406] = 0, [2407] = 0, [2408] = 61025, [2409] = 63215, [2410] = 32767...}
#10 0x00000000010d13dd in S4Importer::ImportSprites (this=0x7fffced23800) at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/S4Importer.cpp:825
No locals.
#11 0x00000000010c2dc4 in S4Importer::Import (this=0x7fffced23800) at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/S4Importer.cpp:148
No locals.
#12 0x00000000010c1218 in rct1_load_saved_game (path=0x283d5e0 <gOpenRCT2StartupActionPath> "/home/janisozaur/Downloads/LUNA PARK by alex.SV4") at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct1/S4Importer.cpp:2229
        result = false
        s4Importer = 0x7fffced23800
#13 0x00000000006a87a7 in rct2_open_file (path=0x283d5e0 <gOpenRCT2StartupActionPath> "/home/janisozaur/Downloads/LUNA PARK by alex.SV4") at /home/janisozaur/workspace/OpenRCT2/src/openrct2/rct2.c:378
        extension = 0x283d60d <gOpenRCT2StartupActionPath+45> "SV4"
#14 0x0000000000edd9e9 in openrct2_launch () at /home/janisozaur/workspace/OpenRCT2/src/openrct2/OpenRCT2.cpp:243
        parkLoaded = false
#15 0x000000000069c045 in main (argc=2, argv=0x7fffffffd538) at /home/janisozaur/workspace/OpenRCT2/src/openrct2/platform/posix.c:59
        run_game = 1
```

**Comment by IntelOrca** [comment]
@janisozaur it failed the assert for me, but I haven't checked every park. Its possible that you get it for vehicles which don't have a 3rd colour - but there is no easy way of knowing which ones do yet.

**Comment by janisozaur** [comment]
If that's the case, can't the colour map be asked if the colour falls outside of it? Or is it not that simple?

**Comment by IntelOrca** [comment]
@janisozaur I think we will have to do that, but its still important we know which vehicles only have two or three colours as we will need to probably set the 3rd colour from a table as part of the conversion to an RCT2 object vehicle

**Comment by janisozaur** [comment]
Requires rebasing after https://github.com/OpenRCT2/OpenRCT2/pull/5019

File: src/openrct2/rct1/S4Importer.cpp [hunk 4]
```diff
@@ -672,6 +672,11 @@ class S4Importer : public IS4Importer
             dst->track_colour_main[0] = RCT1::GetColour(src->track_primary_colour);
             dst->track_colour_additional[0] = RCT1::GetColour(src->track_secondary_colour);
             dst->track_colour_supports[0] = RCT1::GetColour(src->track_support_colour);
+
+            // Balloons were always blue in the original RCT.
+            if (src->type == RCT1_RIDE_TYPE_BALLOON_STALL) {
+                dst->track_colour_main[0] = COLOUR_LIGHT_BLUE;
+            }
```
Comment (IntelOrca) [inline src/openrct2/rct1/S4Importer.cpp]: Brace on its own line

**Comment by Gymnasiast** [comment]
I fixed the formatting and pushed the commit to develop.


By the way, I have yet to encounter a vehicle that actually has a third colour in RCT1. There also doesn't seem to be room for it in the rct1_ride structure. So I think it's probably either padding or another field.

**Comment by IntelOrca** [comment]
> There also doesn't seem to be room for it in the rct1_ride structure

What do you mean?

**Comment by Gymnasiast** [comment]
There seems to be no field in the rct1_ride for vehicle_colours_extended. rct_ride does have that field.

**Comment by IntelOrca** [comment]
Oh fair enough

File: src/openrct2/rct1.h [hunk 5]
```diff
@@ -26,8 +26,9 @@
 #include "world/map.h"
 #include "world/sprite.h"
 
-#define RCT1_MAX_MAP_ELEMENTS   0xC000
-#define RCT1_MAX_SPRITES        5000
+#define RCT1_MAX_MAP_ELEMENTS       0xC000
```
Comment (marijnvdwerf) [inline src/openrct2/rct1.h]: Why are these aligned using spaces? Only seem to be two cases of spaces used in this file

File: src/openrct2/rct1/tables.cpp [hunk 6]
```diff
@@ -1362,7 +1362,7 @@ namespace RCT1
 
 #pragma warning(push)
 #pragma warning(disable : 4505) // 'identifier' unreferenced local function has been removed
```
Comment (marijnvdwerf) [inline src/openrct2/rct1/tables.cpp]: @IntelOrca If this function is made non-static, can't this warning be removed?

File: src/openrct2/rct1/S4Importer.cpp [hunk 7]
```diff
@@ -1178,7 +1176,20 @@ class S4Importer final : public IS4Importer
...
+            const char * peepName = GetUserString(src->name_string_idx);
+            if (peepName[0] != 0)
+            {
+                rct_string_id peepNameStringId = user_string_allocate(4, peepName);
+                if (peepNameStringId != 0)
+                {
+                    dst->name_string_idx = peepNameStringId;
+                }
+            }
+        }
```
Comment (IntelOrca) [inline src/openrct2/rct1/S4Importer.cpp]: This could have been done on `develop`

**Comment by Gymnasiast** [comment]
@IntelOrca I pushed the peep name fix to develop.

File: src/openrct2/rct1.h [hunk 8]
```diff
@@ -1145,6 +1145,8 @@ enum {
 	STANDUP_TWISTER_RC_CARRIAGE = 72,
 	REVERSER_RC_CAR = 79,
 	REVERSER_RC_BOGIE = 80,
+    MINIGOLF_PLAYER = 81,
+    MINIGOLF_BALL = 82,
```
Comment (IntelOrca) [inline src/openrct2/rct1.h]: spaces -> tabs

**Comment by Gymnasiast** [comment]
@IntelOrca I'm not quite sure what causes AppVeyor to trip up. Could you look into it?

**Comment by IntelOrca** [comment]
@Gymnasiast `rct1\S4Importer.cpp(895): warning C4310: cast truncates constant value`

**Comment by janisozaur** [comment]
@IntelOrca the issue is… this is line 895:
```
 ImportPeeps();
```

**Comment by IntelOrca** [comment]
I don't why AppVeyor is saying that, when you check it out the warning is on line 910

**Comment by IntelOrca** [comment]
AppVeyor CI seems to have compiled 8537c8dd, so its correct in that respect:
https://github.com/OpenRCT2/OpenRCT2/blob/8537c8dd/src/openrct2/rct1/S4Importer.cpp#L895

**Comment by Gymnasiast** [comment]
Well, Ted put line 910 there right at the beginning, so I don't know why it only started to complain now. And I'm not sure what to do with it.

**Comment by IntelOrca** [comment]
Replace `(sint16)0x8000` with `LOCATION_NULL` - I can't remember the exact constant name.

File: src/openrct2/rct1/S4Importer.cpp [hunk 9]
```diff
@@ -766,6 +751,109 @@ class S4Importer final : public IS4Importer
...
+
+        if (_gameVersion < FILE_VERSION_RCT1_LL && dst->type == RIDE_TYPE_MERRY_GO_ROUND)
+        {
+            // The merry-go-round in pre-LL versions was always yellow with red
+            dst->vehicle_colours[0].body_colour = COLOUR_YELLOW;
+            dst->vehicle_colours[0].trim_colour = COLOUR_BRIGHT_RED;
+        }
+        else
+        {
+            for (int i = 0; i < 12; i++)
```
Comment (marijnvdwerf) [inline src/openrct2/rct1/S4Importer.cpp]: `RCT1_MAX_VEHICLES_PER_RIDE`?

File: src/openrct2/rct1/S4Importer.cpp [hunk 10]
```diff
@@ -791,37 +879,278 @@ class S4Importer final : public IS4Importer
...
+            FixVehicleLinks(vehicle, spriteIndexMap);
+        }
+        FixRideVehicleLinks(spriteIndexMap);
+    }
+
+    void ImportVehicle(rct_vehicle * dst, rct1_vehicle * src)
+    {
+        char buf[64];
+        sprintf(buf, "%d", src->vehicle_type);
+        log_error(buf);
```
Comment (marijnvdwerf) [inline src/openrct2/rct1/S4Importer.cpp]: Isn't `log_error` a `printf`-like function itself?

Comment (Gymnasiast) [inline src/openrct2/rct1/S4Importer.cpp]: Possibly, but this part is supposed to be removed before merging anyway.

**Comment by Gymnasiast** [comment]
This branch is getting into a state where we can consider merging it.

There are some bugs:
- Non-ascii characters are not imported correctly. 'Märchenfahrt' is changed into 'Mä²£henfahrt'.
- The mechanics are in the wrong place - try the above mentioned 'Märchenfahrt' in Heide-Park for example.
- The wrong mechanic is assigned to a ride.
- Many peeps leave the queue as soon as they can board, though only the first time after a scenario is imported.
- The number of block sections is calculated incorrectly. This error was there before the peep import.

**Comment by IntelOrca** [comment]
> Non-ascii characters are not imported correctly. 'Märchenfahrt' is changed into 'Mä²£henfahrt'.

Which park has this?

**Pull request merged**